common/argv.c

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/internal/argv.h>
+#include <openenclave/internal/raise.h>
+
+#if defined(OE_BUILD_ENCLAVE)
+#include <openenclave/corelibc/string.h>
+#define strlen oe_strlen
+#else
+#include <string.h>
+#endif
+
+#define MAX_LENGTHS 32
+
+oe_result_t oe_argv_to_buffer(
+    const char* argv[],
+    size_t argc,
+    void* buf_out,
+    size_t buf_size,
+    size_t* buf_size_out)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    size_t required_size = 0;
+    size_t i;
+    size_t lengths[MAX_LENGTHS];
+
+    if (!argv || !buf_size_out)
+        goto done;
+
+    /* Handle empty argv list case up front. */
+    if (argc == 0)
+    {
+        *buf_size_out = 0;
+        result = OE_OK;
+        goto done;
+    }
+
+    /* Determine the total memory requirements. */
+    for (i = 0; i < argc; i++)
+    {
+        size_t len;
+
+        if (!argv[i])
+            OE_RAISE(OE_FAILURE);
+
+        len = strlen(argv[i]);
+
+        if (i < MAX_LENGTHS)
+            lengths[i] = len;
+
+        required_size += len + 1;
+    }
+
+    /* Fail if the buffer is too small. */
+    if (buf_size < required_size)
+    {
+        *buf_size_out = required_size;
+        OE_RAISE(OE_BUFFER_TOO_SMALL);
+    }
+
+    /* Copy the strings onto the allocated buffer. */
+    if (buf_out)
+    {
+        char* p = (char*)buf_out;
+
+        for (i = 0; i < argc; i++)
+        {
+            size_t len = (i < MAX_LENGTHS) ? lengths[i] : strlen(argv[i]);
+            memcpy(p, argv[i], len + 1);
+            p += len + 1;
+        }
+    }
+
+    *buf_size_out = required_size;
+    result = OE_OK;
+
+done:
+
+    return result;
+}
+
+oe_result_t oe_buffer_to_argv(
+    const void* buf,
+    size_t buf_size,
+    char*** argv_out,
+    size_t argc,
+    void* (*malloc_func)(size_t),
+    void (*free_func)(void*))
+{
+    oe_result_t result = OE_UNEXPECTED;
+    char** argv = NULL;
+    size_t argv_size;
+    size_t alloc_size = 0;
+
+    if (!buf || !argv_out || !malloc_func || !free_func)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Calculate the total size of argv. */
+    {
+        const char* p = (const char*)buf;
```

Comment:
Please use the safemath.h methods for pointer arithmetic; there's not a strong binding of edger8r validated inputs with this method which could end up being used outside that context.

Reply:
Resolved.